Fix Wash regressions

MikaelSmith opened the pull request:
Fix regressions introduced by #635. Supports #297.

MikaelSmith committed:
Fix regressions introduced by process group and setsid manipulation

Only in release installs of Wash, I encountered a race calling `setsid`
that led to
```
Error starting new session for Wash daemon: operation not permitted
```

Also, running Wash non-interactively would fail
```
Error getting process group of new shell: no such process
Couldn't stop child shell: os: process already finished
```

Found that the important part of fixing the original issue was to put
the Wash daemon in a new session (note that this call only succeeds when
Wash is not the leader of its process group). Simplified the approach to
just call `setsid` after starting the child shell (so that it remains in
the original session).

Also introduces the BATS testing framework.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

Clarify explanation of how we're using process groups and sessions

Signed-off-by: Michael Smith <michael.smith@puppet.com>

ekinanp commented on cmd/rootMain.go:
```diff
@@ -125,24 +125,41 @@ Try 'help'`)
+		close(exitCodeCh)
 	}()
 
+	// Only switch process group if interactive, where the child starts a new session.
```
Nit: This should probably contain some details about why we're doing this for future maintainers. Thinking of something like

> Only switch process group if interactive, where the child starts a new session. This way, the daemon and the shell share the same TTY, which enables plugins to prompt the user for input like e.g. security tokens.

ekinanp commented on README.md:
```diff
@@ -26,6 +26,8 @@ To build it, run `go build`. To test, run `go test ./...`.
 
 > Requires golang 1.12+.
 
+Additional unit tests are in the [test](test) directory. First built with `go build`, then run them using [BATS](https://github.com/bats-core/bats-core) with `PATH=.:$PATH bats test`.
```
This is a pretty good idea. Only concern is that the stuff in `test` tests the Wash binary, so they're not really unit tests. They seem more like acceptance tests?

MikaelSmith replied:
Ah yes, wrote the wrong thing.

MikaelSmith commented:
There's one more issue I'm investigating with zsh
```
zsh: can't set tty pgrp: operation not permitted
```

One difference I've found is that running `go run wash.go` starts with Wash in the `go run ...` process group, and when zsh exits it returns the tty to that process group. Where-as starting `wash`, it starts in its own process group which gets reparented to be under the zsh session, so trying to return the tty to the parent process group fails because you're not allowed to switch control that direction.

MikaelSmith committed:
Ensure the original process group is preserved for zsh

If Wash starts as the leader of its own process group, then when we move
the Wash daemon to a new session we change the permission model of that
original process group.  Some shells, zsh in particular, try to return
the tty to the parent process group when they exit. When that happens
and the parent's process group is now under in a different session, this
fails with
```
zsh: can't set tty pgrp: operation not permitted
```

When Wash detects it starts as the leader of its process group, fork so
that the daemon is not the leader so we create a new process group when
reparenting it under the shell.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
Updated.

ekinanp commented on cmd/rootMain.go:
```diff
@@ -59,6 +59,25 @@ func rootMain(cmd *cobra.Command, args []string) exitCode {
 
+	// If interactive and this process is in its own process group, then fork and run the original
+	// command as a new process that's not the leader of its process group. This is specifically to
+	// work with shells that try to use the parent's process group, which we change later on. By
```
What does "try to use the parent's process group" mean? That some shells try to use the parent's process group's TTY?

MikaelSmith replied:
In this case I think it's returning the TTY to the parent process group. But I don't fully understand what zsh is doing.

ekinanp replied:
Ok. Now say `P` is the server process in pgrp `A`, `C` is the child shell process in pgrp `B`, and that `C` needs `A` (it is one of those shells you mention in the comment). Without this code, we'd get `P -- B` at the end, which resulted a failure because it didn't match `A`. Does that mean that `C` knows it needs `A` when it boots up (i.e. zsh statically determines the parent's pgrp instead of checking it via a system call)?

It's OK if the answer to this is "I have no idea." Just trying to understand it.

MikaelSmith replied:
I suspect it's actually: `C` asks for its parent process id (`P`) and requests the process group (`A`) for that. It then calls `tcsetpgrp` or something similar in `ioctl` (https://github.com/zsh-users/zsh/blob/ad9f07e66fded8a44adba15d576960cef587f9d4/Src/utils.c#L4643-L4674) to transfer the TTY to that process group. Originally that would've worked because `pgrp must be a (nonempty) process group belonging to the same session as the calling process` (https://linux.die.net/man/3/tcsetpgrp), but now `A` is actually a different process group that's no longer part of the same session as `C` (I think `setsid` does that).

Re-reading https://blog.nelhage.com/2010/01/a-brief-introduction-to-termios-signaling-and-job-control/, it's possible that there's an additional step I need to take after moving the daemon to acquire a controlling terminal.

MikaelSmith replied:
I'm going to keep digging into this, because I think there is a little more setup missing. For example
```
USER           PID  PPID  PGID   SESS JOBC STAT   TT       TIME COMMAND
michaelsmith 25735 25734 25735      0    0 Ss     ??    0:04.33 wash
michaelsmith 25167 25166 25167      0    1 S    s001    0:00.86 -zsh
michaelsmith 25734 25167 25734      0    1 S    s001    0:00.09 wash
michaelsmith 25745 25735 25745      0    0 S+   s001    0:00.18 /bin/zsh
michaelsmith 26275 26274 26275      0    1 S    s003    0:00.71 -zsh
michaelsmith 26436 26275 26436      0    1 S    s003    0:00.12 zsh
michaelsmith 26987 26436 26987      0    1 S+   s003    0:00.11 zsh
```

This shows two process trees:
- Run `wash`: 25167 (login zsh) -> 25734 (wash) -> 25735 (sub-wash) -> 25745 (/bin/zsh)
- Run `zsh`, then `zsh` again: 26275 (login zsh) -> 26436 (zsh) -> 26987 (zsh)

Digging into `ps` (on macOS), here are some notable traits:
- everything in a login shell shares the same TTY (except sub-wash, which doesn't seem to know its TTY)
- everything has a session of 0. This seems weird, I think ps on macOS is just not capturing anything useful there.
- `/bin/zsh` has a job control count of 0; seems like it should be 1
- sub-wash is a session leader (the 2nd lower-case `s` under STAT).

MikaelSmith replied:
Figured out session ID, `ps` on macOS doesn't really query it right. Wrote a small C program that checks it
```
printf("pid: %d, process groupd id: %d, session ID: %d\n", pid, getpgid(pid), getsid(pid));
```
and by default everything shares a single session.

What I've done in this PR seems to work the best, haven't come up with any variations. I'll try to wrap up all the explanation in some code comments.

ekinanp replied:
Re: https://github.com/puppetlabs/wash/pull/637#discussion_r354017719, you're saying that the `operation not permitted` error that you found on zsh happened because of something like:

* `P` and `C` share the same session, let's say `S1`
* `P` changes its process group to `B`
* `P` changes its session to `S2`
* `C` asks for `P`'s process group, which is `B`.
* `C` attempts to transfer the TTY to `B`, but that doesn't work because `P`'s session is now `S2` instead of `S1`

MikaelSmith replied:
Yes. Previously they were all in the same process group, now the Wash daemon is in a different process group.

I need to specifically check that this works on Linux as well, because it's getting into some low-level classification. I haven't figured out how to have an acceptance test that exercises the interactive code.

MikaelSmith replied:
Works on Linux.

MikaelSmith replied:
Ok, the new session was the only important thing. So did find a way to simplify it, and I think the explanation is clearer now too.

ekinanp reviewed (approved):
The comments are not blockers.

MikaelSmith commented on cmd/rootMain.go:
```diff
@@ -21,7 +21,16 @@ import (
 var rootCommandFlag string
 
-// Start the wash server, then present the default system shell.
+// Start the Wash server then present the default system shell. The server will be running in the
```
Yeah. I think I'd accidentally run my installed `wash` when I thought it worked.

MikaelSmith commented on cmd/rootMain.go:
```diff
@@ -21,7 +21,16 @@ import (
+// able to prompt for input without interupting the shell (as the controller of the terminal) so
+// that plugins can prompt the user for input like e.g. security tokens.
+//
+// To allow prompts, we start the shell process then put the Wash server (daemon) in a new session.
```
Ok, will update.

ekinanp commented on cmd/rootMain.go:
```diff
@@ -117,79 +145,34 @@ Try 'help'`)
+	// If interactive (when we might prompt the user for input, such as security tokens), create a
+	// new session. If not interactive, calling setsid is pointless and might fail.
+	if plugin.IsInteractive() {
+		if _, err := unix.Setsid(); err != nil {
```
This is a lot simper.

MikaelSmith replied:
Only worked after I started ensuring the daemon wasn't leader of a process group. Put that note in the commit message.

ekinanp commented on cmd/rootMain.go:
```diff
@@ -59,6 +59,26 @@ func rootMain(cmd *cobra.Command, args []string) exitCode {
+	// work with shells (zsh) that try to use their parent's process group, which would no longer be
+	// in the same session. By forking, we keep the original process group in its original session so
+	// the child shell can still modify it.
+	if plugin.IsInteractive() && os.Getpid() == unix.Getpgrp() {
```
From what I'm seeing in https://linux.die.net/man/2/setsid, could the reasoning behind this be simplified to "we ensure that the original command is not the leader of its process group so that `setsid` (which creates the new session) works"?

MikaelSmith replied:
Kind of, bet technically the other approach would have also worked (and we wouldn't need to exec).
